base/strings/string.jl

```diff
@@ -124,13 +124,13 @@ end
 _memcmp(a::Union{Ptr{UInt8},AbstractString}, b::Union{Ptr{UInt8},AbstractString}, len) =
     ccall(:memcmp, Cint, (Ptr{UInt8}, Ptr{UInt8}, Csize_t), a, b, len % Csize_t) % Int
 
-function cmp(a::String, b::String)
+@assume_effects :foldable :removable function cmp(a::String, b::String)
     al, bl = sizeof(a), sizeof(b)
     c = _memcmp(a, b, min(al,bl))
     return c < 0 ? -1 : c > 0 ? +1 : cmp(al,bl)
 end
 
-function ==(a::String, b::String)
+@assume_effects :foldable :removable function ==(a::String, b::String)
```

Comment:
Can we be more specific in these effects or adjust effects on one of the called functions instead?

Reply:
I believe we can't. `_memcmp` takes the size which would lead to UB if it were passed an incorrect length. (I think `pointer_from_objref` is also fundamentally non-consistent).